cuML/src/glm/qn/qn.h

```diff
@@ -42,61 +42,63 @@ int qn_fit(LossFunction &loss, T *Xptr, T *yptr, T *zptr, int N,
   if (l2 == 0) {
     GLMWithData<T, LossFunction> lossWith(&loss, Xptr, yptr, zptr, N, ordX);
 
-    return qn_minimize(w, fx, num_iters, lossWith, l1, opt_param, stream,
-                        verbosity);
+    return qn_minimize(cuml, w, fx, num_iters, lossWith, l1, opt_param, stream,
+                       verbosity);
 
   } else {
 
     Tikhonov<T> reg(l2);
     RegularizedGLM<T, LossFunction, decltype(reg)> obj(&loss, &reg);
     GLMWithData<T, decltype(obj)> lossWith(&obj, Xptr, yptr, zptr, N, ordX);
 
-    return qn_minimize(w, fx, num_iters, lossWith, l1, opt_param, stream,
-                        verbosity);
+    return qn_minimize(cuml, w, fx, num_iters, lossWith, l1, opt_param, stream,
+                       verbosity);
   }
 }
 
 template <typename T>
-void qnFit(T *X, T *y, int N, int D, int C, bool fit_intercept, T l1, T l2,
-           int max_iter, T grad_tol, int linesearch_max_iter, int lbfgs_memory,
-           int verbosity, T *w0, T *f, int *num_iters, bool X_col_major,
-           int loss_type, cublasHandle_t cublas, cudaStream_t stream) {
-
-  // TODO this will come from the cuml handle
+void qnFit(const cumlHandle_impl &cuml, T *X, T *y, int N, int D, int C,
+           bool fit_intercept, T l1, T l2, int max_iter, T grad_tol,
+           int linesearch_max_iter, int lbfgs_memory, int verbosity, T *w0,
+           T *f, int *num_iters, bool X_col_major, int loss_type,
+           cudaStream_t stream) {
 
   STORAGE_ORDER ord = X_col_major ? COL_MAJOR : ROW_MAJOR;
 
-  SimpleMatOwning<T> z(C, N); // TODO this allocation somewhere else?
+  MLCommon::device_buffer<T> tmp(cuml.getDeviceAllocator(), stream, C * N);
+
+  SimpleMat<T> z(tmp.data(), C, N);
```

Comment:
@teju85, @akkamesh Seems to be another occurrence of a tensor view class.

Reply:
Good point. 
We talked about this a while ago and decided to prioritize the features we are working on over unifying our abstractions, which also seemed non-trivial because they were designed with different use cases in mind.

I can imagine though that there might soon be a critical mass of code that could benefit from such abstractions. E.g. in the case of GLM, having the matrix concept made it relatively painless to transition to sparse data support.